File: src/shortcuts.ts
export const shortcutActions = [
  { id: "previous", label: "上一曲", hint: "回到上一首音乐" },
  { id: "next", label: "下一曲", hint: "继续下一首音乐" },
  { id: "toggle", label: "播放 / 暂停", hint: "随时停下，随时继续" },
  { id: "volumeUp", label: "音量增加", hint: "播放器音量 +5%" },
  { id: "volumeDown", label: "音量减少", hint: "播放器音量 −5%" },
] as const;
export type ShortcutAction = (typeof shortcutActions)[number]["id"];
export type ShortcutConfig = {
  enabled: boolean;
  bindings: Record<ShortcutAction, string>;
};
export type ShortcutStatus = {
  portal: boolean;
  triggers: Partial<Record<ShortcutAction, string>>;
};
export const shortcutStorageKey = "lori-global-shortcuts";
export const defaultShortcuts: ShortcutConfig = {
  enabled: false,
  bindings: {
    previous: "Ctrl+Alt+ArrowLeft",
    next: "Ctrl+Alt+ArrowRight",
    toggle: "Ctrl+Alt+Space",
    volumeUp: "Ctrl+Alt+ArrowUp",
    volumeDown: "Ctrl+Alt+ArrowDown",
  },
};

const allowedKey =
  /^(Key[A-Z]|Digit[0-9]|F([1-9]|1[0-9]|2[0-4])|Arrow(Left|Right|Up|Down)|Space|Home|End|PageUp|PageDown|Insert|Delete|Backspace|Enter)$/;
export function validShortcut(value: string) {
  if (!value) return true;
  const parts = value.split("+");
  const key = parts.pop()!;
  return (
    allowedKey.test(key) &&
    parts.length > 0 &&
    parts.some((p) => p !== "Shift") &&
    parts.every((p) => ["Ctrl", "Alt", "Shift", "Super"].includes(p)) &&
    new Set(parts).size === parts.length
  );
}
export function parseShortcutConfig(raw: string | null): ShortcutConfig {
  if (!raw) return structuredClone(defaultShortcuts);
  const value = JSON.parse(raw);
  if (
    typeof value?.enabled !== "boolean" ||
    !value.bindings ||
    shortcutActions.some(
      ({ id }) =>
        typeof value.bindings[id] !== "string" ||
        !validShortcut(value.bindings[id]),
    )
  ) {
    throw new Error("已保存的快捷键格式无效，请在设置中重新保存。");
  }
  if (shortcutConflict(value))
    throw new Error("已保存的快捷键存在重复，请在设置中重新保存。");
  return {
    enabled: value.enabled,
    bindings: Object.fromEntries(
      shortcutActions.map(({ id }) => [id, value.bindings[id]]),
    ) as ShortcutConfig["bindings"],
  };
}
export function shortcutConflict(config: ShortcutConfig) {
  const seen = new Set<string>();
  for (const { id, label } of shortcutActions) {
    const key = config.bindings[id];
    if (!key) continue;
    const canonical = key.split("+").sort().join("+");
    if (seen.has(canonical))
      return `${label}与其他操作使用了相同快捷键，请更换组合。`;
    seen.add(canonical);
  }
  return "";
}
export function recordShortcut(
  event: Pick<
    KeyboardEvent,
    | "code"
    | "ctrlKey"
    | "altKey"
    | "shiftKey"
    | "metaKey"
    | "repeat"
    | "isComposing"
  >,
): string | null {
  if (event.repeat || event.isComposing || !allowedKey.test(event.code))
    return null;
  if (!event.ctrlKey && !event.altKey && !event.metaKey) return null;
  return [
    event.ctrlKey && "Ctrl",
    event.altKey && "Alt",
    event.shiftKey && "Shift",
    event.metaKey && "Super",
    event.code,
  ]
    .filter(Boolean)
    .join("+");
}
export function shortcutLabel(value: string) {
  return value
    .split("+")
    .map(
      (key) =>
        ({
          ArrowLeft: "←",
          ArrowRight: "→",
          ArrowUp: "↑",
          ArrowDown: "↓",
          Space: "空格",
          Super: "Win / Super",
        })[key] ?? key.replace(/^(Key|Digit)/, ""),
    )
    .join(" + ");
}
export function adjustedVolume(current: number, direction: number) {
  return Math.max(
    0,
    Math.min(1, Math.round((current + direction * 0.05) * 100) / 100),
  );
}

File: src-tauri/src/shortcuts.rs
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc, RwLock,
};
use tauri::{Emitter, Manager};
use tauri_plugin_global_shortcut::{GlobalShortcutExt, Shortcut, ShortcutState};

const ACTIONS: [(&str, &str); 5] = [
    ("previous", "上一曲"),
    ("next", "下一曲"),
    ("toggle", "播放 / 暂停"),
    ("volumeUp", "音量增加"),
    ("volumeDown", "音量减少"),
];

#[derive(Clone, Deserialize)]
pub struct Config {
    enabled: bool,
    bindings: HashMap<String, String>,
}

#[derive(Default)]
pub struct Shortcuts {
    update: tokio::sync::Mutex<()>,
    active: Arc<RwLock<HashMap<u32, String>>>,
    native_keys: std::sync::Mutex<Vec<Shortcut>>,
    #[cfg(target_os = "linux")]
    portal: tokio::sync::Mutex<Option<portal::Registration>>,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Status {
    portal: bool,
    triggers: HashMap<String, String>,
}

#[tauri::command]
pub fn shortcut_backend() -> &'static str {
    if uses_portal() {
        "portal"
    } else {
        "native"
    }
}

pub fn uses_portal() -> bool {
    cfg!(target_os = "linux")
        && (std::env::var_os("WAYLAND_DISPLAY").is_some()
            || std::env::var("XDG_SESSION_TYPE").is_ok_and(|v| v == "wayland"))
}

fn validate(config: &Config) -> Result<Vec<(String, String, Shortcut)>, String> {
    let mut seen = HashSet::new();
    let mut result = Vec::new();
    if config
        .bindings
        .keys()
        .any(|key| !ACTIONS.iter().any(|(id, _)| id == key))
    {
        return Err("未知快捷键操作".into());
    }
    for (action, label) in ACTIONS {
        let value = config
            .bindings
            .get(action)
            .map(String::as_str)
            .unwrap_or("");
        if value.is_empty() {
            continue;
        }
        let shortcut = value
            .parse::<Shortcut>()
            .map_err(|e| format!("{label}：{e}"))?;
        if !seen.insert(shortcut.id()) {
            return Err(format!("{label}与其他操作使用了相同快捷键"));
        }
        result.push((action.into(), value.into(), shortcut));
    }
    Ok(result)
}

#[tauri::command]
pub async fn apply_shortcuts(app: tauri::AppHandle, config: Config) -> Result<Status, String> {
    let mut desired = validate(&config)?;
    if !config.enabled {
        desired.clear();
    }
    let state = app.state::<Shortcuts>();
    let _guard = state.update.lock().await;
    #[cfg(target_os = "linux")]
    if uses_portal() {
        return portal::apply(&app, &state, &desired).await;
    }
    let old = state.active.read().map_err(|e| e.to_string())?.clone();
    let mut added = Vec::new();
    // Reserve new keys first: a conflict never removes the working bindings.
    for (_, text, key) in &desired {
        if old.contains_key(&key.id()) {
            continue;
        }
        let actions = state.active.clone();
        let pressed = Arc::new(AtomicBool::new(false));
        if let Err(error) = app
            .global_shortcut()
            .on_shortcut(*key, move |app, key, event| {
                if event.state == ShortcutState::Released {
                    pressed.store(false, Ordering::SeqCst);
                } else if !pressed.swap(true, Ordering::SeqCst) {
                    if let Ok(actions) = actions.read() {
                        if let Some(action) = actions.get(&key.id()) {
                            let _ = app.emit_to("main", "lori-shortcut", action);
                        }
                    }
                }
            })
        {
            let mut cleanup = Vec::new();
            for key in added {
                if let Err(e) = app.global_shortcut().unregister(key) {
                    cleanup.push(e.to_string());
                }
            }
            return Err(format!(
                "无法注册 {text}，可能已被系统或其他应用占用：{error} {}",
                cleanup.join("；")
            ));
        }
        added.push(*key);
    }
    let next: HashMap<_, _> = desired
        .iter()
        .map(|(action, _, key)| (key.id(), action.clone()))
        .collect();
    // Removing a key only affects our own registrations.
    // Resolve the old IDs via the saved native shortcut objects.
    let removal = {
        let mut keys = state.native_keys.lock().map_err(|e| e.to_string())?;
        let mut errors = Vec::new();
        for key in keys.iter().filter(|key| !next.contains_key(&key.id())) {
            if let Err(error) = app.global_shortcut().unregister(*key) {
                errors.push(error.to_string());
            }
        }
        *keys = desired.iter().map(|(_, _, key)| *key).collect();
        errors
    };
    *state.active.write().map_err(|e| e.to_string())? = next;
    if !removal.is_empty() {
        return Err(format!("快捷键释放失败：{}", removal.join("；")));
    }
    Ok(Status {
        portal: false,
        triggers: desired.into_iter().map(|(a, t, _)| (a, t)).collect(),
    })
}

#[cfg(target_os = "linux")]
mod portal {
    use super::*;
    use ashpd::desktop::{
        global_shortcuts::{GlobalShortcuts, NewShortcut},
        Session,
    };
    use futures_util::StreamExt;

    pub struct Registration {
        session: Session<GlobalShortcuts>,
        activated: tauri::async_runtime::JoinHandle<()>,
        changed: tauri::async_runtime::JoinHandle<()>,
        enabled: Arc<AtomicBool>,
    }

    fn triggers(
        shortcuts: &[ashpd::desktop::global_shortcuts::Shortcut],
    ) -> HashMap<String, String> {
        shortcuts
            .iter()
            .map(|s| {
                (
                    s.id().split("--").next().unwrap_or("").into(),
                    s.trigger_description().into(),
                )
            })
            .collect()
    }

    pub async fn apply(
        app: &tauri::AppHandle,
        state: &Shortcuts,
        desired: &[(String, String, Shortcut)],
    ) -> Result<Status, String> {
        let mut current = state.portal.lock().await;
        if desired.is_empty() {
            if let Some(old) = current.as_ref() {
                old.session.close().await.map_err(|e| e.to_string())?;
            }
            if let Some(old) = current.take() {
                old.enabled.store(false, Ordering::SeqCst);
                old.activated.abort();
                old.changed.abort();
            }
            return Ok(Status {
                portal: true,
                triggers: HashMap::new(),
            });
        }
        let portal = GlobalShortcuts::new()
            .await
            .map_err(|e| format!("桌面未提供全局快捷键接口：{e}"))?;
        let session = portal
            .create_session(Default::default())
            .await
            .map_err(|e| e.to_string())?;
        // Include the preferred key in the stable ID. Desktops may remember a
        // binding by ID; changing a key in Lori must request a new binding.
        let shortcuts: Vec<_> = desired
            .iter()
            .map(|(action, key, _)| {
                let label = ACTIONS.iter().find(|(id, _)| id == action).unwrap().1;
                NewShortcut::new(format!("{action}--{key}"), format!("Lori · {label}"))
                    .preferred_trigger(portal_trigger(key).as_str())
            })
            .collect();
        let prepared = async {
            let activated = portal.receive_activated().await?;
            let changed = portal.receive_shortcuts_changed().await?;
            let response = portal
                .bind_shortcuts(&session, &shortcuts, None, Default::default())
                .await?
                .response()?;
            Ok::<_, ashpd::Error>((activated, changed, triggers(response.shortcuts())))
        }
        .await;
        let (mut activated, mut changed, triggers) = match prepared {
            Ok(value) => value,
            Err(error) => {
                let _ = session.close().await;
                return Err(format!("未启用全局快捷键，桌面授权可能已取消：{error}"));
            }
        };
        if let Some(old) = current.as_ref() {
            if let Err(error) = old.session.close().await {
                let _ = session.close().await;
                return Err(format!("无法释放原快捷键：{error}"));
            }
        }
        if let Some(old) = current.take() {
            old.enabled.store(false, Ordering::SeqCst);
            old.activated.abort();
            old.changed.abort();
        }
        let enabled = Arc::new(AtomicBool::new(true));
        let active = enabled.clone();
        let path = serde_json::to_value(&session)
            .map_err(|e| e.to_string())?
            .as_str()
            .ok_or("无效的桌面会话")?
            .to_owned();
        let handle = app.clone();
        let activated_task = tauri::async_runtime::spawn(async move {
            while let Some(event) = activated.next().await {
                if event.session_handle().as_str() == path && active.load(Ordering::SeqCst) {
                    let action = event.shortcut_id().split("--").next().unwrap_or("");
                    if ACTIONS.iter().any(|(id, _)| *id == action) {
                        let _ = handle.emit_to("main", "lori-shortcut", action);
                    }
                }
            }
        });
        let path = serde_json::to_value(&session)
            .map_err(|e| e.to_string())?
            .as_str()
            .ok_or("无效的桌面会话")?
            .to_owned();
        let handle = app.clone();
        let active = enabled.clone();
        let changed_task = tauri::async_runtime::spawn(async move {
            while let Some(event) = changed.next().await {
                if event.session_handle().as_str() == path && active.load(Ordering::SeqCst) {
                    let _ = handle.emit_to(
                        "main",
                        "lori-shortcuts-changed",
                        super::Status {
                            portal: true,
                            triggers: self::triggers(event.shortcuts()),
                        },
                    );
                }
            }
        });
        *current = Some(Registration {
            session,
            activated: activated_task,
            changed: changed_task,
            enabled,
        });
        Ok(Status {
            portal: true,
            triggers,
        })
    }
}

#[cfg(any(target_os = "linux", test))]
fn portal_trigger(key: &str) -> String {
    key.split('+')
        .map(|part| match part {
            "Ctrl" => "CTRL",
            "Alt" => "ALT",
            "Shift" => "SHIFT",
            "Super" => "LOGO",
            "ArrowLeft" => "Left",
            "ArrowRight" => "Right",
            "ArrowUp" => "Up",
            "ArrowDown" => "Down",
            "Space" => "space",
            other => other
                .strip_prefix("Key")
                .or_else(|| other.strip_prefix("Digit"))
                .unwrap_or(other),
        })
        .collect::<Vec<_>>()
        .join("+")
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn rejects_duplicates_and_unknown_actions_before_registration() {
        let mut config = Config {
            enabled: true,
            bindings: HashMap::from([
                ("next".into(), "Ctrl+Alt+ArrowRight".into()),
                ("previous".into(), "Ctrl+Alt+ArrowRight".into()),
            ]),
        };
        assert!(validate(&config).is_err());
        config.bindings.remove("previous");
        assert_eq!(validate(&config).unwrap().len(), 1);
        config.bindings.insert("bad".into(), "Space".into());
        assert!(validate(&config).is_err());
    }
    #[test]
    fn converts_recorded_keys_to_portal_accelerators() {
        assert_eq!(portal_trigger("Ctrl+Alt+ArrowLeft"), "CTRL+ALT+Left");
        assert_eq!(portal_trigger("Super+Shift+KeyP"), "LOGO+SHIFT+P");
        assert_eq!(portal_trigger("Ctrl+Alt+Space"), "CTRL+ALT+space");
        assert_eq!(portal_trigger("Ctrl+Digit9"), "CTRL+9");
    }
}
